### PR TITLE
feat: custom binding names for out of runtime deployment

### LINKED ISSUE #4093: ✨[Feature] Support `dict[str, torch.Tensor]` inputs
**Is your feature request related to a problem? Please describe.**
It's a fairly common pattern to have a dictionary of tensors as the input.

e.g.
```python
class MyModule(torch.nn.Module):
    def forward(self, inputs: dict[str, torch.Tensor]) -> dict[str, torch.Tensor]:
        ...

```

**Describe the solution you'd like**
Currently the inputs are expected to be `torch.Tensor | list[torch.Tensor]`, I would like this to be extended to `torch.Tensor | list[torch.Tensor] | dict[str, torch.Tensor] | list[dict[str, torch.Tensor]]`

**Describe alternatives you've considered**

I have attempted to use hacks manipulating the signature of the `forward` function to change the names of the input arguments.

**Additional context**

### LINKED ISSUE #4212: ❓ [Question] How can you set the output names when converting to serialized engine?
## ❓ Question

I am converting my model to a serialized engine using (model is a `torch.nn.Module` instance)
```
program = torch_tensorrt.dynamo.trace(model, arg_inputs=example_inputs)
engine_bytes: bytes = torch_tensorrt.dynamo.convert_exported_program_to_serialized_trt_engine(
        program,
        arg_inputs=example_inputs,
        require_full_compilation=True,
        immutable_weights=True,
        tiling_optimization_level="full",
        use_python_runtime=False,
        enabled_precisions={torch.float16, torch.bfloat16, torch.float32},
        use_explicit_typing=False,
        optimization_level=5,
        version_compatible=True,
        hardware_compatible=True,
)
```
The input names of the serialized engine are named after the inputs of the nn.Module, but the output names are just named output0, output1 ...

How can I set these names to something more useful? I need certain names for the model to work in our production environment.

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Allows users of the `convert_exported_program_to_serialized_engine` API to specify binding names for **all** bindings utilizing the exported programs pytrees for i/o flattening to flatten the provided names and associate names to specific io positions. 

TODO: Add support for aliasing inputs and outputs through this method (see: #4251)

Fixes #4212

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Review by lanluo-nvidia — APPROVED** [review]
LGTM
